Fix windows libc++ `fs::path` `fstream` compile errors

ryanofsky opened the pull request:
Drop support for passing `fs::path` directly to `std::ifstream` and `std::ofstream` constructors and `open()` functions, because as reported by hebasto in https://github.com/bitcoin/bitcoin/issues/33545, after https://wg21.link/lwg3430 there is no way this can continue to work in windows builds, and there are already compile errors compiling for windows with newer versions of libc++.

Instead, add an `fs::path::std_path()` method that returns `std::filesystem::path` references and use it where needed.

ryanofsky committed:
Fix windows libc++ fs::path fstream compile errors

As reported by hebasto in https://github.com/bitcoin/bitcoin/issues/33545,
newer libc++ versions implementing https://wg21.link/lwg3430 will no longer
implicitly convert `fs::path` objects to `std::filesystem::path` objects when
constructing `std::ifstream` and `std::ofstream` types.

This is not a problem in Unix systems since `fs::path` objects use
`std::string` as their native string type, but it causes compile errors on
Windows which use `std::wstring` as their string type, since `fstream`s can't
be constructed from `wstring`s.

Fix the windows libc++ compile errors by adding a new `fs::path::std_path()`
method and using it construct `fstream`s more portably.

Additionally, delete `fs::path`'s implicit `native_string` conversion so these
errors will not go undetected in the future, even though there is not currently
a CI job testing Windows libc++ builds.

hebasto reviewed (commented):
Concept ACK.

Maybe consider removing these lines now:https://github.com/bitcoin/bitcoin/blob/f39d3e1a2f2b9cdcf0e77942fd69eb92a58ce77f/src/util/fs.h#L79-L81 ?

ryanofsky committed:
Simplify fs::path by dropping filename() and make_preferred() overloads

These overloads were needed to allow passing `fs::path` objects directly to
libstdc++'s `fstream` constructors, but after the previous commit, there is no
longer any remaining code that does pass `fs::path` objects to `fstream`
constructors. Writing new code which does this is also discouraged because the
standard has been updated in https://wg21.link/lwg3430 to disallow it.

Dropping these also means its no longer possible to pass `fs::path` arguments
directly to `fstream::open` in libstdc++, which is somewhat unfortunate but not
a big loss because it is already not possible to pass them to the constructor.
So this commit updates `fstream::open` calls.

Additionally, this change required updates to src/bitcoin.cpp since it was
relying on the overloaded filename() method.

hebasto reviewed (commented):
Tested b0113afd44b4c7c0d0da9883424bd2978de3d18c by cross-compiling for `x86_64-w64-mingw32` and `aarch64-w64-mingw32` targets using the [LLVM MinGW](https://github.com/mstorsjo/llvm-mingw) toolchain.

hebasto reviewed (approved):
ACK c864a4c1940d682f7eb6fdb3b91b18d638b59330.

hebasto reviewed (commented):
> ... even though there is not currently a CI job testing Windows libc++ builds.

Added to the nightly builds in https://github.com/hebasto/bitcoin-core-nightly/pull/74.

fanquake commented:
> Added to the nightly builds in https://github.com/hebasto/bitcoin-core-nightly/pull/74.

Looks like there are failing unit tests (https://github.com/hebasto/bitcoin-core-nightly/actions/runs/18313967330/job/52151177201#step:6:401):
```bash
test\util_string_tests.cpp(43): Entering test case "ConstevalFormatString_NumSpec"
test/util_string_tests.cpp(40): error: in "util_string_tests/ConstevalFormatString_NumSpec": exception "const char*" raised as expected: validation on the raised exception through predicate "HasReason{error}"
test/util_string_tests.cpp(40): error: in "util_string_tests/ConstevalFormatString_NumSpec": exception "const char*" raised as expected: validation on the raised exception through predicate "HasReason{error}"
...
```

hebasto commented:
> > Added to the nightly builds in [hebasto/bitcoin-core-nightly#74](https://github.com/hebasto/bitcoin-core-nightly/pull/74).
> 
> Looks like there are failing unit tests (https://github.com/hebasto/bitcoin-core-nightly/actions/runs/18313967330/job/52151177201#step:6:401):
> 
> ```shell
> test\util_string_tests.cpp(43): Entering test case "ConstevalFormatString_NumSpec"
> test/util_string_tests.cpp(40): error: in "util_string_tests/ConstevalFormatString_NumSpec": exception "const char*" raised as expected: validation on the raised exception through predicate "HasReason{error}"
> test/util_string_tests.cpp(40): error: in "util_string_tests/ConstevalFormatString_NumSpec": exception "const char*" raised as expected: validation on the raised exception through predicate "HasReason{error}"
> ...
> ```

The issue appears to be caused by [this](https://github.com/bitcoin/bitcoin/commit/6c3c619b35cc03e883f9d2b3326f906aedde9ba7) commit.

Friendly ping @l0rinc @hodlinator as co-authors of that commit :)

hebasto commented:
I've submitted an [issue](https://github.com/mstorsjo/llvm-mingw/issues/522) to https://github.com/mstorsjo/llvm-mingw.

@l0rinc 

> I have reproduced the error in [l0rinc/bitcoin-core-nightly#2](https://github.com/l0rinc/bitcoin-core-nightly/pull/2) with @hebasto's nightly clone redirected to `l0rinc/master`.
> 
> I pushed a potential fix that adds a `const char *` overload to `HasReason` in [l0rinc#43](https://github.com/l0rinc/bitcoin/pull/43) - tested through [l0rinc/bitcoin-core-nightly#1](https://github.com/l0rinc/bitcoin-core-nightly/pull/1).
> 
> Also added an alternative which replaces all remaining `const char *` throws with `std::runtime_error` in [l0rinc#44](https://github.com/l0rinc/bitcoin/pull/44) - tested through [l0rinc/bitcoin-core-nightly#3](https://github.com/l0rinc/bitcoin-core-nightly/pull/3).
> 
> Based on your feedback and on what the CI gods think, I will push one of them as a PR.

I believe that [this](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#e15-throw-by-value-catch-exceptions-from-a-hierarchy-by-reference) is also a good point. Because of [array-to-pointer](https://en.cppreference.com/w/cpp/language/implicit_cast.html#Array-to-pointer_conversion) conversion, I'd lean to the second option you mentioned.

---

> Also note that the windows build indicates additional warnings (unused vars because of `ifndef __USING_WINDOWS__`)
>
> We should likely fix these as well - I don't mind pushing a separate PR for this after the other one.

Yes. These warnings are unrelated to the current PR changes, and I'd prefer to see them fixed in separated PR(s).

l0rinc commented:
> I'd lean to the second option you mentioned

Thanks, closed the other ones and opened https://github.com/bitcoin/bitcoin/pull/33569 - the dedicated nightly build passed: https://github.com/l0rinc/bitcoin-core-nightly/pull/3

l0rinc reviewed (approved):
Code review ACK c864a4c1940d682f7eb6fdb3b91b18d638b59330

It bothers me a bit that the first commit doesn't compile, but it seems fixed in the second one - I'm fine with merging as long as the second commit is part of the PR.

I have tried reproducing the error of removing `make_preferred` and `filename` on `master`, but I'm not actually getting any failure - but it doesn't look dangerous, I'm personally fine with removing those, they don't seem to be necessary anymore.

<details>
<summary>Details</summary>

```
root@i9-ssd /mnt/my_storage/bitcoin # cat /usr/include/c++/15/fstream | grep -C2 make_preferred
  // Enable if _Path is a filesystem::path or experimental::filesystem::path
  template<typename _Path, typename _Result = _Path, typename _Path2
           = decltype(std::declval<_Path&>().make_preferred().filename())>
    using _If_fs_path = enable_if_t<is_same_v<_Path, _Path2>, _Result>;
#endif // C++17
root@i9-ssd /mnt/my_storage/bitcoin # git log -1
commit 0eeae4d174a41c3fc2eae41e76b929fa3114d6f3 (HEAD, origin/master, origin/HEAD)
Merge: d30f149360 879c21045e
Author: merge-script <fanquake@gmail.com>
Date:   Sun Oct 19 15:45:47 2025 +0100

    Merge bitcoin/bitcoin#33625: Update secp256k1 subtree to latest master

    3cbf7cb3e6ac51492b354732bddbb4f58ce97ed3 Squashed 'src/secp256k1/' changes from b9313c6e1a..d543c0d917 (fanquake)

    Pull request description:

      Updates the subtree to https://github.com/bitcoin-core/secp256k1/commit/d543c0d917a76a201578948701cc30ef336e0fe6
      Related to #33284.

    ACKs for top commit:
      hebasto:
        ACK 879c21045eba64b3dc875f6f2c2c579abecde1d0.
      janb84:
        ACK 879c21045eba64b3dc875f6f2c2c579abecde1d0

    Tree-SHA512: 1802cd84959b5c935170792f458651f30431fe8340ead7966ff381c1c0c3a9f6c21bbb8dd96a07482ffed49642ded49e80b61802e688b8351956b111dffd5a78
root@i9-ssd /mnt/my_storage/bitcoin # git diff
diff --git a/src/util/fs.h b/src/util/fs.h
index 7c313e2072..24f4943048 100644
--- a/src/util/fs.h
+++ b/src/util/fs.h
@@ -65,11 +65,6 @@ public:
         const std::u8string& utf8_str{std::filesystem::path::u8string()};
         return std::string{utf8_str.begin(), utf8_str.end()};
     }
-
-    // Required for path overloads in <fstream>.
-    // See https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=96e0367ead5d8dcac3bec2865582e76e2fbab190
-    path& make_preferred() { std::filesystem::path::make_preferred(); return *this; }
-    path filename() const { return std::filesystem::path::filename(); }
 };

 static inline path u8path(const std::string& utf8_str)
 ```

</details>

I have also tried reverting `std::ofstream fout{file};` and the deleted method does indeed catch it, so the implementation seems correct regardless of whether it fixes the issue - which I have reproduced with @hebasto's nightly branch anyway.

I have also tested why the `auto` was needed in `wrapper_dir` - it seems to be related to mixing paths for the `/` - seems fine to me.

ryanofsky commented:
Thanks for the review!

> It bothers me a bit that the first commit doesn't compile

It should compile. Github just seems to show a red x next to the commit because the job was cancelled when the second commit was pushed.

> I have tried reproducing the error of removing make_preferred and filename on master, but I'm not actually getting any failure

I think the failure only happened on windows. The workaround was introduced here: https://github.com/bitcoin/bitcoin/pull/20744#issuecomment-1028380153

maflcko reviewed (commented):
re-ACK c864a4c1940d682f7eb6fdb3b91b18d638b59330 🌥

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: re-ACK c864a4c1940d682f7eb6fdb3b91b18d638b59330 🌥
0D0v6O8aoNqJXCgjkkFNkvQkFzEvahdV0ADufk2lsrbmdkWx9WCf6o0r/LORbJuEwFehUzATQ8meSoax1jCVBw==
```

</details>
